sound/soc/sof/amd/pci-rn.c

```diff
@@ -0,0 +1,162 @@
+
+#define ACP3x_REG_START		0x1240000
+#define ACP3x_REG_END		0x125C000
+
+static struct platform_device *dmic_dev;
+static struct platform_device *pdev;
+
+static struct resource renoir_res[] = {
+	{
+		.start = 0,
+		.end = ACP3x_REG_END - ACP3x_REG_START,
+		.name = "acp_mem",
+		.flags = IORESOURCE_MEM,
+	},
+	{
+		.start = 0,
+		.end = 0,
+		.name = "acp_dai_irq",
+		.flags = IORESOURCE_IRQ,
+	},
+};
+
+static const struct sof_dev_desc renoir_desc = {
+	.machines		= snd_soc_acpi_amd_sof_machines,
+	.resindex_lpe_base	= 0,
+	.resindex_pcicfg_base	= -1,
+	.resindex_imr_base	= -1,
+	.irqindex_host_ipc	= -1,
+	.default_fw_path	= "amd/sof",
+	.default_tplg_path	= "amd/sof-tplg",
+	.default_fw_filename	= "sof-rn.ri",
+	.nocodec_tplg_filename	= "sof-acp.tplg",
+	.ops			= &sof_renoir_ops,
+};
+
+static int acp_pci_rn_probe(struct pci_dev *pci, const struct pci_device_id *pci_id)
+{
+	struct device *dev = &pci->dev;
+	struct resource *res;
+	struct resource res_i2s;
+	struct platform_device_info pdevinfo;
+	unsigned int flag, i, addr;
+	int num_res, ret;
+
+	num_res = ARRAY_SIZE(renoir_res);
+
+	flag = snd_amd_acp_find_config(pci);
+	if (flag != FLAG_SOF && flag != FLAG_SOF_ONLY_DMIC)
+		return -ENODEV;
+
+	ret = sof_pci_probe(pci, pci_id);
+	if (ret != 0)
+		return -EINVAL;
+
+	if (!dmic_dev) {
+		dmic_dev = platform_device_register_data(dev, "dmic-codec", PLATFORM_DEVID_NONE,
+							 NULL, 0);
+		if (IS_ERR(dmic_dev)) {
+			dev_err(dev, "error: failed to create DMIC device\n");
+			sof_pci_remove(pci);
+			return PTR_ERR(dmic_dev);
+		}
+	}
+
+	/*
+	 * We have common PCI driver probe for ACP device but we have to support I2S without SOF
+	 * for some machine. Register platform device for I2S use case without SOF if machine flag
+	 * is set to FLAG_SOF_ONLY_DMIC.
+	 */
+
+	if (flag == FLAG_SOF_ONLY_DMIC && !pdev) {
+		addr = pci_resource_start(pci, 0);
+		res = devm_kzalloc(&pci->dev, sizeof(struct resource) * num_res, GFP_KERNEL);
+		if (!res) {
+			sof_pci_remove(pci);
+			return -ENOMEM;
+		}
+
+		for (i = 0 ; i < num_res; i++) {
+			res_i2s = renoir_res[i];
+			res[i].name = res_i2s.name;
+			res[i].flags = res_i2s.flags;
+			res[i].start = addr + res_i2s.start;
+			res[i].end = addr + res_i2s.end;
+			if (res_i2s.flags == IORESOURCE_IRQ) {
+				res[i].start = pci->irq;
+				res[i].end = res[i].start;
+			}
+
+			res_i2s = res[i];
+		}
+		memset(&pdevinfo, 0, sizeof(pdevinfo));
+
+		pdevinfo.name = "acp_asoc_renoir";
+		pdevinfo.id = 0;
+		pdevinfo.parent = &pci->dev;
+		pdevinfo.num_res = ARRAY_SIZE(renoir_res);
+		pdevinfo.res = &res[0];
+
+		pdev = platform_device_register_full(&pdevinfo);
+		if (IS_ERR(pdev)) {
+			dev_err(&pci->dev, "cannot register %s device\n",
+				pdevinfo.name);
+			ret = PTR_ERR(pdev);
+		}
+	}
+
+	return ret;
+};
+
+static void acp_pci_rn_remove(struct pci_dev *pci)
+{
+	if (!IS_ERR_OR_NULL(dmic_dev))
+		platform_device_unregister(dmic_dev);
+	if (!IS_ERR_OR_NULL(pdev))
+		platform_device_unregister(pdev);
```

Comment:
IIRC the IS_ERR_OR_NULL is no longer necessary. We haven't removed this from the Intel code but now I think it's not needed. Please double-check.